### PR TITLE
ByteArrayStreamInput: Return -1 when there are no more bytes to read

### PR DESCRIPTION
**Pull request by chrisberkhout**
Returning `-1` avoids readers getting stuck in a loop.

This is the same logic used for [single-byte reads](https://github.com/elastic/elasticsearch/blob/eaf27bdda38d3a94bdf099ac8e1f0ba0feab72da/server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java#L44-L50).

Without that [`java.awt.Font::createFont0`](https://github.com/openjdk/jdk17/blob/master/src/java.desktop/share/classes/java/awt/Font.java#L1153) was getting stuck in a loop, when used from [`‎org.elasticsearch.xpack.esql.expression.function.RailRoadDiagram::loadFont`](https://github.com/elastic/elasticsearch/blob/44758b3823a4982e9c62598f184cf90b67ab5d2a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/RailRoadDiagram.java#L203).

---

 - Relates #112072

### CODE REVIEW

**Review by nik9000 — DISMISSED** [review]
LGTM. Given that this is shared I think it should probably get a unit test. If it were just for the silly font thing it'd be fine, but someday someone's going to rely on this.

**Comment by DaveCTurner** [comment]
++ yeah I think this needs something in `org.elasticsearch.common.io.stream.AbstractStreamTests`, that way we check this behaviour on all (present and future) impls.

File: server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java [hunk 1]
```diff
@@ -120,6 +120,9 @@ public void readBytes(byte[] b, int offset, int len) {
 
     @Override
     public int read(byte[] b, int off, int len) throws IOException {
+        if (available() <= 0) {
+            return -1;
+        }
         int toRead = Math.min(len, available());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/io/stream/ByteArrayStreamInput.java:126]: Also let's avoid calling `available()` twice, this is on a hot enough path for that to matter I think

**Review by DaveCTurner — APPROVED** [review]
LGTM thanks @chrisberkhout!

**Pull request merged**